### PR TITLE
Upgrade to Kafka 3.7.0

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

_Select the type of your PR_

- Enhancement / new feature

### Description

Upgrading to Kafka 3.7.0

### Additional Context

_Why are you making this pull request?_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [x] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

**Comment by k-wall** [comment]
@strimzi-ci run system tests

**Comment by k-wall** [comment]
@franvila FYI

**Comment by k-wall** [comment]
> ❗ **System tests Failed (no tests results are present)** ❗

the system test run is unable to find the Kafka 3.7.0 tgz.

**Comment by franvila** [comment]
@strimzi-ci run system tests

**Comment by k-wall** [comment]
We discussed the system test failure. It is because the code is using the kafka artefact version number to drive the kafka version requested in the Kafka CR.  The version of Strimzi doesn't know about 3.7.0, so hence the fail.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/codec/KafkaRequestDecoder.java
Comment (robobario) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/codec/KafkaRequestDecoder.java:147]: Should we change this to `>= 9`? Or is it something we need to check against each new apiVersion?

Comment (k-wall) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/codec/KafkaRequestDecoder.java:147]: I seem to remember an earlier conversation where we agreed we wanted the prompt to check the schema.

File: pom.xml
Comment (k-wall) [inline pom.xml:739]: Don't merge me!

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/filter/BrokerAddressFilter.java
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/filter/BrokerAddressFilter.java:122]:
```suggestion
        // KIP-951, Version 16
```

**Comment by k-wall** [comment]
@franvila @SamBarker  even once Apache Kafka 3.7 is out, we still won't be able to merge this PR until there's a version of Strimzi out that supports Kafka 3.7. 

I think it would be advantageous to relax the dependency.  It would mean we could see the remainder of the test suite pass against 3.7 early.  That would help de-risk the 0.5.0 release.

If we were to define a temporary kafka.version=3.6.1 property in the system test pom, I believe that would allow the system tests to run against Kafka 3.6.1, whilst the rest of the testsuite would use Kafka 3.7.0.   Once Strimzi is released upstream, we raise a PR to update the strimzi dependency and the kafka.version override could be removed.

Or maybe you see a better way?

**Comment by k-wall** [comment]
@strimzi-ci run system tests

**Comment by franvila** [comment]
> @franvila @SamBarker even once Apache Kafka 3.7 is out, we still won't be able to merge this PR until there's a version of Strimzi out that supports Kafka 3.7.
> 
> I think it would be advantageous to relax the dependency. It would mean we could see the remainder of the test suite pass against 3.7 early. That would help de-risk the 0.5.0 release.
> 
> If we were to define a temporary kafka.version=3.6.1 property in the system test pom, I believe that would allow the system tests to run against Kafka 3.6.1, whilst the rest of the testsuite would use Kafka 3.7.0. Once Strimzi is released upstream, we raise a PR to update the strimzi dependency and the kafka.version override could be removed.
> 
> Or maybe you see a better way?

In System tests, to deploy kafka we are using Strimzi, so we need to wait for it to support 3.7.0. Test clients need to be updated as well to support kafka 3.7.0. So I cannot see the benefit of making the system tests pass forcing an older kafka version in the PR. We will get a pass but not with the version we want to test (false positive).

Currently passing the system tests are not mandatory in a PR, so if you like to tests against kafka 3.6.1, there is another job in Jenkins to launch manually the tests once the PR is merged to main. There you can select the kafka version to be used, so you can perform that test, but I wouldn't recommend to create temporary things that could be forgotten.

**Comment by k-wall** [comment]
> Currently passing the system tests are not mandatory in a PR, so if you like to tests against kafka 3.6.1, there is another job in Jenkins to launch manually the tests once the PR is merged to main. There you can select the kafka version to be used, so you can perform that test, but I wouldn't recommend to create temporary things that could be forgotten.

But it'd be weird to merge a change that would break the developer's normal workflow.  It means that any developer on the Kroxylicious project would need to remember to run the system tests in a special way (assuming they have the access).  The overhead of that seems much greater than the overhead of updating two system properties in the system tests POM.

I agree we'd need a process to make sure we remember to update both system properties.  Wouldn't just a comment next to the strimzi.version property suffice?  We can even put the two properties together.

**Comment by franvila** [comment]
> > Currently passing the system tests are not mandatory in a PR, so if you like to tests against kafka 3.6.1, there is another job in Jenkins to launch manually the tests once the PR is merged to main. There you can select the kafka version to be used, so you can perform that test, but I wouldn't recommend to create temporary things that could be forgotten.
> 
> But it'd be weird to merge a change that would break the developer's normal workflow. It means that any developer on the Kroxylicious project would need to remember to run the system tests in a special way (assuming they have the access). The overhead of that seems much greater than the overhead of updating two system properties in the system tests POM.
> 
> I agree we'd need a process to make sure we remember to update both system properties. Wouldn't just a comment next to the strimzi.version property suffice? We can even put the two properties together.

Sorry Keith, but from a testing point of view I don't still see that solution. We cannot say that we are passing the system tests when it is not true. False positives are really dangerous, because we can think that everything is fine when it may not.

I'd prefer to say that we cannot run the system tests because of that reason, instead of running against something it does not correspond and can cause a false confidence.

File: pom.xml
Comment (k-wall) [reply pom.xml]: The maven artefacts are on Maven Central, staging repo removed.

File: kroxylicious-systemtests/pom.xml
Comment (franvila) [inline kroxylicious-systemtests/pom.xml:27]: I'm not comfortable having a different version here. Having a pass in the system tests does not mean that it will work fine for kafka 3.7.0, so from testing point of view it does not make any sense to give that result.

**Comment by k-wall** [comment]
It is in no way a false positive - the system tests are passing under a known/defined configuration.  

Let me restate the problem I'm trying solve.

* I don't want to block taking Kafka 3.7 release until Strimzi 0.30 is released.  By taking the Kafka 3.7 release now, we  help derisk - have more time for problems to emerge and more time to find solutions.
* I don't really want to prevent work other work going on in the system test module in the meanwhile.  For instance, there can be work going on towards #918 which doesn't need to be impeded.  If the system tests are in a broken state for n days.
* Test suites being in a broken state is just a bad thing.  It is too easy for developers to introduce regressions and the suite gets more broken, without anyone realising.

We know we have to upgrade Strimzi before 0.5.0 already, so I don't really see any way it could get forgotten.

**Comment by franvila** [comment]
> It is in no way a false positive - the system tests are passing under a known/defined configuration.

Yes, we have a known config, but in the PR pipeline we won't be testing the proper version of kafka.

I understand your point and I don't want to stop the development as well. I just want to warn that it could cause some not found bugs during the period Strimzi is releasing the new version. If we want to assume that risk, because it can cause more benefits than damages, we can go ahead.

**Comment by k-wall** [comment]
thanks @franvila, let's hear from the others.

**Comment by k-wall** [comment]
just finding out if my idea actually flies or not :)

@strimzi-ci run system tests

**Comment by franvila** [comment]
@strimzi-ci run system tests

**Comment by SamBarker** [comment]
I agree with 
> I don't want to block taking Kafka 3.7 release until Strimzi 0.30 is released. By taking the Kafka 3.7 release now, we help derisk - have more time for problems to emerge and more time to find solutions.

We don't want to be beholden to Strimzi as to when we change Kafka versions. 

That said I also largely agree with @franvila that defining a different version for the system tests to run against gives a false sense of confidence that everything is working. In the general case I don't think it is is material which Kafka version we test against but for upgrade PR's its quite problematic. 

The question is how do we resolve those tensions? 

I think the short answer is we need to be able to have Strimzi deploy our Kafka versions. Even if that means we need to build Strimzi ourselves (we just need to to deploy a Kafka instance we are not interested in testing Strimzi support for the new Kafka release), but more likely we just need to override the Kafka Images using https://strimzi.io/docs/operators/0.39.0/full/configuring#con-common-configuration-images-reference and possibly the version properties... 


FYI the Strimzi release we are likely to need is `0.40.0` not `0.30.0` as mentioned in the quote but that has no bearing on the discussion.

File: kroxylicious-krpc-plugin/src/test/resources/Data/example-expected-FetchRequest.java.txt
Comment (SamBarker) [inline kroxylicious-krpc-plugin/src/test/resources/Data/example-expected-FetchRequest.java.txt:111]: surely this should return the `HIGHEST_SUPPORTED_VERSION` constant and not hard code its own value

**Review by SamBarker — COMMENTED** [review]
I'm happy with the changes outside of the system tests but I don't feel entirely comfortable approving without the system tests having run.

File: kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/filter/BrokerAddressFilterTest.java
Comment (SamBarker) [reply kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/filter/BrokerAddressFilterTest.java:154]: So its clear, I'm not requesting or suggesting changes here but curious to hear your thoughts on the options.

**Comment by k-wall** [comment]
> I think the short answer is we need to be able to have Strimzi deploy our Kafka versions

The Kafka image you refer won't exist until the Strimzi release is made.  AFAIK Strimzi won't operate with any old Kafka image (the new Apache Kafka image, the Kafka native image, bitnami...).

**Comment by k-wall** [comment]
I see the consensus is that we want to hold this PR until the Strimzi release is public.

**Review by SamBarker — APPROVED** [review]
As mentioned in the call lets merge as is.

File: kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/filter/BrokerAddressFilterTest.java
Comment (k-wall) [reply kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/filter/BrokerAddressFilterTest.java:154]: I'd seen us use this approach elsewhere in the project so I went with it.

I think the alternative you refer to is:

```
when(myMock.myFunction(anyString())).thenAnswer(i -> i.getArguments()[0]);
```

which has the advantage of less captor clutter, but as you say, at the expense of type safety.

I'd value type safety more. I think as the project grows having to deal with a bunch of runtime failures scattered across tests (after a refactoring) would be harder work than deal with compile time failures.

**Pull request merged**
